src/main/java/org/jetbrains/research/kotlincodesmelldetector/utils/FirUtils.kt

```diff
@@ -0,0 +1,87 @@
+package org.jetbrains.research.kotlincodesmelldetector.utils
+
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.fir.declarations.FirFile
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.fir.declarations.FirVariable
+import org.jetbrains.kotlin.fir.java.FirProjectSessionProvider
+import org.jetbrains.kotlin.fir.psi
+import org.jetbrains.kotlin.fir.resolve.dfa.FirControlFlowGraphReferenceImpl
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.*
+import org.jetbrains.kotlin.fir.resolve.transformers.FirTotalResolveTransformer
+import org.jetbrains.kotlin.idea.fir.FirModuleResolveStateImpl
+import org.jetbrains.kotlin.idea.fir.getOrBuildFir
+import org.jetbrains.kotlin.utils.DFS
+
+fun extractFirSimpleFunctions(firFile: FirFile): List<FirSimpleFunction> {
+    val result = mutableListOf<FirSimpleFunction>()
+    result.addAll(firFile.declarations.filterIsInstance<FirSimpleFunction>())
+    // TODO other types of classes
+    result.addAll(firFile.declarations.filterIsInstance<FirRegularClass>().flatMap {
+        it.declarations.filterIsInstance<FirSimpleFunction>()
+    })
+    return result
+}
+
+fun getCurrentFirFileOpenInEditor(project: Project): FirFile {
+    val ktFile = getCurrentFileOpenInEditor(project)
+    val transformer = FirTotalResolveTransformer()
+    val firFile = ktFile?.getOrBuildFir(FirModuleResolveStateImpl(FirProjectSessionProvider(project)))!!
+    transformer.processFiles(listOf(firFile))
```

Comment:
`getOrBuildFir` already runs transformers accordingly to given resolve phase

Reply:
Functions have empty cfg references without calling `transformer.processFiles`

Reply:
That's (probably) because default resolve phase in `getOrBuildFir` is `FirResolvePhase.DECLARATIONS` (analyze declarations but not their bodies), and you need `FirResolvePhase.BODY_RESOLVE` (analyze declarations with bodies).